Geometry cleanups and modernization

nyalldawson opened the pull request:
Following on from comments in https://github.com/qgis/QGIS/pull/5349, this PR cleans up internal memory management in QgsGeometry and related classes by moving everything possible to unique_ptrs and making cloning/non-cloning detachments more explicit.

nyalldawson committed:
Rework internal geometry memory management

Use unique_ptrs wherever possible, and cleanup geometry
detachments

nyalldawson committed:
Cleanup some loops

nyalldawson committed:
Fix travis build

nyalldawson committed:
Move qgsAsConst to qgis::as_const

So it's more inline with the std::as_const implementation which
it fills in for, and allows us to 'polyfill' other c++>11
features into the qgis:: namespace.

nyalldawson committed:
'Polyfill' std::make_unique to qgis::make_unique

nyalldawson committed:
Even more memory safety in geometry internals by using qgis::make_unique

nyalldawson committed:
Some safer memory management for geos

nyalldawson committed:
Safer memory management in geos

Also speed up avoid intersections by removing unnecessary geometry
cloning

nyalldawson commented:
@wonder-sk look ok to you?

wonder-sk commented:
I am +0 on this one... just because I'm an old-fashioned person who's scared of learning semantics of the fancy new c++ stuff like make_unique / std::move etc. ;-)

One suggestion regarding detachAndClone() and detachAndReset(geom) - how about calling them just detach() and reset(geom) ? I think with reset it is implied that it will detach - or the detachAndSomething() pattern could extend to others methods that detach as well - which would not be great :-)

nyalldawson commented:
> One suggestion regarding detachAndClone() and detachAndReset(geom) - how about calling them just detach() and reset(geom) 

Good idea - i'll do that

>I am +0 on this one... just because I'm an old-fashioned person who's scared of learning semantics of the fancy new c++ stuff like make_unique / std::move etc. ;-)

@wonder-sk and c++11

![200w_d](https://user-images.githubusercontent.com/1829991/31539766-fead6210-b04c-11e7-812c-3717a17a2841.gif) ?

nyalldawson committed:
Rename detach methods in QgsGeometry for clarity

Now we have:

- detach (which clones the existing geometry)
and
- reset (which detaches without cloning and sets a new geometry)
